### PR TITLE
[#IC-419] Upgrade to fp-ts@2

### PR DESCRIPTION
**Pull request by balanza**

### CODE REVIEW

File: src/utils/config.ts [hunk 1]
```diff
@@ -222,36 +222,38 @@ const errorOrConfig: t.Validation<IConfig> = IConfig.decode({
...
       flow(
         IntegerFromString.decode,
         E.mapLeft(() => DEFAULT_SERVER_PORT)
       )
     ),
-    E.fold(identity, identity)
+    E.toUnion
   ),
   TOKEN_EXPIRATION: pipe(
+    // FIXME: if env var is empty string, the result of the pipe would be 0.
```
Comment (balanza) [inline src/utils/config.ts:255]: @AleDore do you think it's a bug?

**Comment by balanza** [comment]
We decided to disable the following `tslint` rules: `prettier` and `ordered-imports` (https://github.com/pagopa/hub-spid-login-ms/pull/49/commits/38db358037c26cbcd8e5deea71d0b6b0eb7af019). That's because they're mainly style rules, the auto-fixer is broken and we're hopefully going to migrate to `eslint` so that would have been wasted work. 

A bunch of lint issues remain, I think they are valid and need to be addressed properly. I'll comment out in code.

File: src/handlers/token.ts [hunk 2]
```diff
@@ -59,189 +68,242 @@ export const getTokenExpiration = (tokenUser: TokenUser | TokenUserL2) =>
...
+              mapDecoding(TokenUserL2, _),
+              TE.mapLeft((e) =>
                 res.status(500).json({
                   detail: String(e),
-                  error: "Error decoding L1 token"
+                  error: "Error decoding L2 token",
                 })
               )
+            )
+      )
```
[the remark this answers is not in the document]
Comment (balanza) [reply src/handlers/token.ts:154]: Beside of that, I'm confused about the whole logic. Now is:
```
IF token is TokenUserL2
THEN decode token into TokenUserL2 (???)
ELSE decode token into TokenUser
```
Why decode something into X if is already X? I see the point is to have type-narrowing, but using a guard (_`.is`_ is a guard actually) is a bit unsafe when dealing with _unsafe_ values.

A more safe approach would be to decode into a _safe_ value; only then, we may use guards to better narrow types:
```
decode token into (TokenUser | TokenUserL2)
IF decode fails 
THEN handle invalid token format
ELSE 
  IF token is TokenUserL2
  THEN ..... 
  ELSE .... (token is TokenUser here)
```

File: src/handlers/token.ts [hunk 3]
```diff
@@ -59,189 +68,242 @@ export const getTokenExpiration = (tokenUser: TokenUser | TokenUserL2) =>
...
       )
-    )
-    .chain(() =>
-      getRawTokenUserFromRedis(req.body.token, res)
-        .chain<TokenUser | TokenUserL2>(_ =>
-          TokenUserL2.is(_)
-            ? mapDecoding(TokenUserL2, _).mapLeft(e =>
+    ),
+    TE.chain<any, boolean, TokenUser | TokenUserL2>(() =>
+      pipe(getRawTokenUserFromRedis(req.body.token, res), (_) =>
```
Comment (balanza) [inline src/handlers/token.ts:134]: What's the type of `_`? 
Screen Shot 2022-04-28 at 12 54 42

I think we're missing a `chain` here.

File: src/handlers/token.ts [hunk 4]
```diff
@@ -59,189 +68,242 @@ export const getTokenExpiration = (tokenUser: TokenUser | TokenUserL2) =>
...
   const fromErrToBadRequest = toBadRequest(res);
-  await fromEither(
-    UpgradeTokenBody.decode(req.body).mapLeft(fromErrToBadRequest)
-  )
-    .chain(_ =>
-      fromEither(NonEmptyString.decode(req.headers[tokenHeaderName])).bimap(
-        e =>
-          fromErrToBadRequest(e, `Missing required header ${tokenHeaderName}`),
-        header => ({
+  pipe(
```
Comment (balanza) [inline src/handlers/token.ts:208]: [Lint error](https://dev.azure.com/pagopaspa/cgn-onboarding-portal-projects/_build/results?buildId=29215&view=logs&j=46d68182-90a1-5651-c618-877729e79262&t=021bd946-7d50-5bb7-93cf-b58765b2698c&l=22):
```
src/handlers/token.ts:208:3 - Promises must be handled appropriately
```

File: src/handlers/token.ts [hunk 5]
```diff
@@ -59,189 +68,242 @@ export const getTokenExpiration = (tokenUser: TokenUser | TokenUserL2) =>
...
         () => res.status(200).json({ active: true })
       )
-    )
-    .chain(() =>
-      getRawTokenUserFromRedis(req.body.token, res)
-        .chain<TokenUser | TokenUserL2>(_ =>
-          TokenUserL2.is(_)
-            ? mapDecoding(TokenUserL2, _).mapLeft(e =>
+    ),
+    TE.chain<any, boolean, TokenUser | TokenUserL2>(() =>
```
[the remark this answers is not in the document]
Comment (LoFrance) [reply src/handlers/token.ts]: Fix with 507781d

File: src/handlers/token.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (balanza) [reply src/handlers/token.ts]: In the end, I decided to rewrite this parte in https://github.com/pagopa/hub-spid-login-ms/pull/49/commits/98e74009555375cf8aa477b3b7062329da9a41c9.

File: src/handlers/token.ts (hunk 3 above)
Comment (balanza) [reply src/handlers/token.ts]: solved by https://github.com/pagopa/hub-spid-login-ms/commit/98e74009555375cf8aa477b3b7062329da9a41c9.

**Comment by iwoak** [comment]
@balanza @LoFrance @AleDore performed manual integration tests both on [dev](https://dev.selfcare.pagopa.it/) using spid-test-env and [prod](https://selfcare.pagopa.it/) by PosteID idp

**Review by iwoak — APPROVED** [review]
@balanza @LoFrance @AleDore performed manual integration tests both on [dev](https://dev.selfcare.pagopa.it/) using spid-test-env and [prod](https://selfcare.pagopa.it/) by PosteID idp.
I used only specific self care configuration: JWT and no ADE api (mock)

**Pull request merged**